programs/leap-util/actions/bls.cpp

```diff
@@ -0,0 +1,125 @@
+
+   // sub-subcommand - key 
+   auto* create_key = create->add_subcommand("key", "Create a new keypair and print the public and private keys")->callback([err_guard]() { err_guard(&bls_actions::create_key); });
+   create_key->add_option("-f,--file", opt->key_file, "Name of file to write private/public key output to. (Must be set, unless \"--to-console\" is passed");
+   create_key->add_flag( "--to-console", opt->print_console, "Print private/public keys to console.");
+
+   // sub-subcommand - pop (proof of possession) 
+   auto* create_pop = create->add_subcommand("pop", "Create proof of possession for the corresponding public key of a given private key")->callback([err_guard]() { err_guard(&bls_actions::create_pop); });
+   create_pop->add_option("-f,--file", opt->key_file, "Name of file storing the private key. (one and only one of \"-f,--file\" and \"--private-key\" must be set)");
+   create_pop->add_option("--private-key", opt->private_key_str, "The private key. (one and only one of \"-f,--file\" and \"--private-key\" must be set)");
+}
+
+int bls_actions::create_key() {
+   if (opt->key_file.empty() && !opt->print_console) {
+      std::cerr << "ERROR: Either indicate a file using \"--file\" or pass \"--to-console\"" << std::endl;
+      return -1;
+   } else if (!opt->key_file.empty() && opt->print_console) {
+      std::cerr << "ERROR: Only one of \"--file\" or pass \"--to-console\" can be provided" << std::endl;
+      return -1;
+   }
+
+   // create a private key and get its corresponding public key
+   const bls_private_key private_key = bls_private_key::generate();
+   const bls_public_key public_key = private_key.get_public_key();
+
+   // generate pop
+   const std::string pop_str = generate_pop_str(private_key);
+
+   // prepare output
+   std::string out_str = "Private key: " + private_key.to_string({}) + "\n";
+   out_str += "Public key: " + public_key.to_string({}) + "\n";
+   out_str += "Proof of Possession: " + pop_str + "\n";
+   if (opt->print_console) {
+      std::cout << out_str;
+   } else {
+      std::cout << "saving keys to " << opt->key_file << std::endl;
+      std::ofstream out( opt->key_file.c_str() );
+      out << out_str;
+   }
+
+   return 0;
+}
+
+int bls_actions::create_pop() {
+   if (opt->key_file.empty() && opt->private_key_str.empty()) {
+      std::cerr << "ERROR: Either indicate a file using \"-f, --file\" or pass \"--private-key\"" << std::endl;
+      return -1;
+   } else if (!opt->key_file.empty() && !opt->private_key_str.empty()) {
+      std::cerr << "ERROR: Only one of \"-f, --file\" and \"--private-key\" can be provided" << std::endl;
+      return -1;
+   }
+
+   std::string private_key_str;
+   if (!opt->private_key_str.empty()) {
+      private_key_str = opt->private_key_str;
+   } else {
+      std::ifstream key_file(opt->key_file);
+
+      if (!key_file.is_open()) {
+         std::cerr << "ERROR: failed to open file " << opt->key_file << std::endl;
+         return -1;
```

Comment:
Do we still want to throw a `CLI::RuntimeError` in those cases?

Reply:
It is thrown in the single place `err_guard`.

Reply:
If we return `-1` here it will throw the exception, right?

Reply:
Yes, it will https://github.com/AntelopeIO/leap/blob/0d585203f95bb227ac6cf8a15bdf362ffb08a416/programs/leap-util/actions/bls.cpp#L18

Reply:
I was not sure that that's we wanted, but OK.